rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/locations/Locations.java

```diff
@@ -35,12 +35,41 @@
 import org.eclipse.lsp4j.TextDocumentItem;
 import org.rascalmpl.uri.URIResolverRegistry;
 import org.rascalmpl.uri.URIUtil;
-import org.rascalmpl.values.IRascalValueFactory;
 
 import io.usethesource.vallang.ISourceLocation;
 import io.usethesource.vallang.IValue;
 
+/**
+ * The Locations class provides utility methods related to source locations and VS Code locations.
+ *
+ * Source locations of files that are opened in the IDE are redirected by prefixing the scheme with "lsp+".
+ * These lsp-redirected source locations must not leak outside of the LSP server. The `toClientLocation` methods
+ * strip source locations of their "lsp+" prefix.
+ */
 public class Locations {
+    public static ISourceLocation toClientLocation(ISourceLocation loc) throws IOException {
+        var result = URIResolverRegistry.getInstance().logicalToPhysical(loc);
+        if (result.getScheme().startsWith("lsp+")) {
+            try {
+                result = URIUtil.changeScheme(result, result.getScheme().substring("lsp+".length()));
```

Comment:
looks like a clone of this extists in `LSPOpenFileResolver`? can we merge these (or perhaps call the function in `LSPOpenFileResolver` here?

Reply:
Good catch!